crates/wasmtime/src/component/types.rs

```diff
@@ -757,3 +759,195 @@ impl Type {
         }
     }
 }
+
+/// Core function type
+#[derive(Debug)]
+pub struct CoreFunc(Handle<SignatureIndex>);
+
+impl CoreFunc {
+    pub(crate) fn from(index: SignatureIndex, ty: &InstanceType<'_>) -> Self {
+        Self(Handle::new(index, ty))
+    }
+
+    /// Returns [`FuncType`] associated with this function
+    pub fn ty(&self) -> FuncType {
+        FuncType::from_wasm_func_type(self.0.types[self.0.index].clone())
+    }
+}
+
+/// Component function type
+#[derive(Debug)]
+pub struct ComponentFunc(Handle<TypeFuncIndex>);
+
+impl ComponentFunc {
+    pub(crate) fn from(index: TypeFuncIndex, ty: &InstanceType<'_>) -> Self {
+        Self(Handle::new(index, ty))
+    }
+
+    /// Iterates over types of function parameters
+    pub fn params(&self) -> impl Iterator<Item = Type> + '_ {
+        let params = self.0.types[self.0.index].params;
+        self.0.types[params]
+            .types
+            .iter()
+            .map(|ty| Type::from(ty, &self.0.instance()))
+    }
+
+    /// Iterates over types of function results
+    pub fn results(&self) -> impl Iterator<Item = Type> + '_ {
+        let results = self.0.types[self.0.index].results;
+        self.0.types[results]
+            .types
+            .iter()
+            .map(|ty| Type::from(ty, &self.0.instance()))
+    }
+}
+
+/// Core module type
+#[derive(Debug)]
+pub struct Module(Handle<TypeModuleIndex>);
+
+impl Module {
+    pub(crate) fn from(index: TypeModuleIndex, ty: &InstanceType<'_>) -> Self {
+        Self(Handle::new(index, ty))
+    }
+
+    /// Iterates over imports of the module
+    pub fn imports(&self) -> impl Iterator<Item = (&(String, String), ExternType)> {
+        self.0.types[self.0.index].imports.iter().map(|(name, ty)| {
+            (
+                name,
+                ExternType::from_wasmtime(self.0.types.module_types(), ty),
+            )
+        })
+    }
+
+    /// Iterates over exports of the module
+    pub fn exports(&self) -> impl Iterator<Item = (&String, ExternType)> {
+        self.0.types[self.0.index].exports.iter().map(|(name, ty)| {
+            (
+                name,
+                ExternType::from_wasmtime(self.0.types.module_types(), ty),
+            )
+        })
+    }
+}
+
+/// Component type
+#[derive(Debug)]
+pub struct Component(Handle<TypeComponentIndex>);
+
+impl Component {
+    pub(crate) fn from(index: TypeComponentIndex, ty: &InstanceType<'_>) -> Self {
+        Self(Handle::new(index, ty))
+    }
+
+    /// Returns import associated with `name`, if such exists in the component
+    pub fn get_import(&self, name: &str) -> Option<ComponentItem> {
+        self.0.types[self.0.index]
+            .imports
+            .get(name)
+            .map(|ty| ComponentItem::from(ty, &self.0.instance()))
+    }
+
+    /// Iterates over imports of the component
+    pub fn imports(&self) -> impl Iterator<Item = (&String, ComponentItem)> {
```

Comment:
Could this get the `ExactSizeIterator` plus `&str` treatment? (and `exports` below too)